Fix google map offsets in china

modos189 opened the pull request:
So, this plugin fixes coordinate correction for some maps in China. #64 

It is also now possible to enable correction not only for Google maps.

modos189 committed:
Modify "China" boundary for offset fix plugin
https://github.com/iitc-project/ingress-intel-total-conversion/pull/1188 #2

modos189 committed:
Update Fix Google Map offsets in China to GoogleMutant. But not working =(
The plugin changes coordinates, but the change is not applied

modos189 committed:
An example of a working solution. Described in more detail in a task
https://github.com/IITC-CE/ingress-intel-total-conversion/issues/64

modos189 committed:
Smooth operation with map, fixed get map options, fixed download of map tiles (except Google maps)

Commit:
fix-googlemap-china-offset: fix style issues

lint:
* `let` is not in es5
* 'HK_LENGTH' is assigned a value but never used
* 'original' is defined but never used

style:
* == vs ===
* Mixed spaces and tabs
* fix indents, missed/extra/trailing spaces, missed semi

modos189 committed:
Fix offset in Google maps

Commit:
Merge branch 'fix-google-map-offsets-in-china' into fix-google-map-offsets-in-china

modos189 committed:
Rename to fix-china-offset.user.js

modos189 committed:
Edited the documentation, because now plugin works not only in Google maps

Commit:
fix-googlemap-china-offset: refactoring

divide PRCoords into 2 distinct functions:
* plugin.fixChinaOffset.isInGoogle
* plugin.fixChinaOffset.gcjObfs

added stub for old fix-googlemap-china-offset plugin

Commit:
fixup! fix-googlemap-china-offset: refactoring

modos189 committed:
Revert changes in Leaflet.GoogleMutant.js

modos189 committed:
Rename fix-china-offset -> fix-china-map-offset

modos189 committed:
Previous indent returned

johnd0e commented:
It it possible to extend L.GridLayer via leaflet api (instead of monkey-patching)?

johnd0e commented on plugins/fix-china-map-offset.js:
```diff
@@ -0,0 +1,313 @@
+  };
+})(L.GridLayer.prototype._setZoomTransform);
+
+L.GridLayer.GoogleMutant.prototype._update = (function (original) {
```
`original` is never used

```suggestion
L.GridLayer.GoogleMutant.prototype._update = (function () {
```

johnd0e commented on plugins/fix-china-map-offset.js:
```diff
@@ -0,0 +1,313 @@
+      var center = this._map.getCenter();
+      /// modified here ///
+      center = window.plugin.fixChinaOffset.getLatLng(center, this.options);
+      /////////////////////
```
In order to make eslint happy
```suggestion
      /////////////////////
      /* eslint-disable */
```

johnd0e commented on plugins/fix-china-map-offset.js:
```diff
@@ -0,0 +1,313 @@
+
+        if (this._mutantIsReady) this._checkZoomLevels();
+        //else zoom level check will be done later by 'idle' handler
+      }
```
```suggestion
      }
      /* eslint-enable */
```

johnd0e commented on plugins/fix-china-map-offset.js:
```diff
@@ -0,0 +1,313 @@
+})(L.GridLayer.GoogleMutant.prototype._update);
+
+window.plugin.fixChinaOffset.getLatLng = function (pos, options) {
+  // No offsets in satellite and hybrid maps
```
It is true for Google map, but it's not reliable to to extrapolate such assumption for arbitrary map.

So I propose to check real maps.
And may be we need additional configuration option.
```suggestion
  // No offsets in satellite and hybrid maps
  // TODO: it's true for Google maps, need to examine other available Chinese maps
```

modos189 replied:
No, this fix is true for all cards that have a correction.

Another thing is that in exceptional cases, when map is connected not by Leaflet, but by means of its own libraries, it may be necessary to correct loading of tiles in them. Only loading, but not coordinates.

johnd0e replied:
> No, this fix is true for all cards that have a correction.

Well, I mean next line:
```
  if (options.needFixChinaOffset === true && options.type !== 'satellite' && options.type !== 'hybrid') {
```
Are you sure that in other maps we also should skip correction for satellite and hybrid?

> when map is connected not by Leaflet

I suppose that it also could be fixed. But let's not run ahead.

johnd0e replied:
https://en.wikipedia.org/wiki/Restrictions_on_geographic_data_in_China:
> Some map providers, such as Here, choose to also offset their satellite imagery layer to match the GCJ-02 street map.[34]

So we need more flexible options.

modos189 commented:
> It it possible to extend L.GridLayer via leaflet api (instead of monkey-patching)?

I don't know, it's the only way I can do(

Commit:
Update plugins/fix-china-map-offset.js

Co-Authored-By: modos189 <yandex@modos189.ru>

Commit:
Update plugins/fix-china-map-offset.js

Co-Authored-By: modos189 <yandex@modos189.ru>

Commit:
Update plugins/fix-china-map-offset.js

Co-Authored-By: modos189 <yandex@modos189.ru>

johnd0e commented:
https://leafletjs.com/examples/extending/extending-1-classes.html
https://leafletjs.com/examples/extending/extending-2-layers.html
It's seems a matter of using of `L.GridLayer.extend(...)` (or sort of).
May be we even could create separate leaflet-plugin, usable in any project, not only iitc).

johnd0e commented:
As we named the plugin `fix-china-map-offset` we should include Baidu-related stuff here.
* Additional transformation needed: https://github.com/Artoria2e5/PRCoords/blob/2193bea83ade11726a61543f548e4f0d9d9304b2/js/PRCoords.js#L141-L154
* Boundaries slightly differ: https://github.com/Artoria2e5/PRCoords/blob/2193bea83ade11726a61543f548e4f0d9d9304b2/js/misc/insane_is_in_china.js#L151-L158

~~So our plugin api gets sort of complex..~~

johnd0e commented:
P.S.
`plugins/fix-china-offset.user.js → plugins/fix-china-map-offset.js`
Need to be ...`user.js`

See #72

Commit:
fix-china-map-offset: Baidu support

* Sources refactored to be closer to upstream https://github.com/Artoria2e5/PRCoords
* Baidu support
* Clarify comment about usage

Commit:
fix-china-map-offset: appied to googleMutant

Commit:
fix-china-map-offset: fix filename

Commit:
fix-china-map-offset: fixme: options for HongKong and map type

Commit:
fixup! fix-china-map-offset: fixme: options for HongKong and map type

johnd0e commented:
>>It it possible to extend L.GridLayer via leaflet api (instead of monkey-patching)?

>I don't know, it's the only way I can do(

Well, we have a couple examples which do not require leaflet monkey-patching:

1. Look here: https://github.com/iitc-project/ingress-intel-total-conversion/pull/1262
It doesn't work in our current environment, but I should be [fixable](https://gist.github.com/johnd0e/4509cd02b71d60510be28ec7ffb8bd38)
(_my fix is quick and dirty, just to make sure if it ever could be done_).
And I suppose `L.GaodeLayer` could be written in more general way, allowing to be reused by other plugins.

2. And here we see another way https://github.com/hobo0cn/leaflet-gcj02
Haven't tried yet, but it seems to be cleaner method.
**Edit**: CRS is property of map, not just TileLayer, so I'm not sure if we can use it..

May be it couldn't be used on google.com, but see: there is google.cn also.

Commit:
fix-china-map-offset: refactor (intermediate function eliminated)

Commit:
fix-china-map-offset: refactor (redefine leaflet methods cleaner)

modos189 commented:
> And here we see another way https://github.com/hobo0cn/leaflet-gcj02
> CRS is property of map, not just TileLayer, so I'm not sure if we can use it..

Yes, this method does not suit us

Commit:
fix-china-map-offset: cosmetics (restore original source indents)

Commit:
fix-china-map-offset: change leaflet extending

do not touch L.GridLayer
extend L.TileLayer and L.GridLayer.GoogleMutant

Commit:
fix-china-map-offset: set 'needFixChinaOffset' option in boot.js

Commit:
fix-china-map-offset: fix: put actions into setup function

Commit:
fix-china-map-offset: patch GoogleMutant in plugin completely (not in boot.js)

Commit:
fixup! fix-china-map-offset: cosmetics (restore original source indents)

Commit:
fix-china-map-offset: remove Baidu-stuff

In current state it doesn't function.
Maybe someone reimplement it in future.
More info here: #64.

johnd0e commented:
New feature idea: instead of hardcoding polygon we could allow import from draw-tools or [geojson](https://gist.github.com/johnd0e/2d61d1fec27da8fc09428fe82b84c3e6#file-china-boundaries-geojson): http://geojson.io/#id=gist:johnd0e/2d61d1fec27da8fc09428fe82b84c3e6&map=4/37.85/104.06

Commit:
fix-china-map-offset: shorten 'namespace' access

Commit:
fix-china-map-offset: fix: L.Util may not be available yet

Commit:
fix-china-map-offset: fix: namespace conflict

Commit:
Merge pull request #72 from johnd0e/fix-google-map-offsets-in-china

update to 'Fix google map offsets in china'

johnd0e commented:
To be **Squash and merge**
